src/nni_manager/training_service/reusable/environments/remoteEnvironmentService.ts

```diff
@@ -0,0 +1,309 @@
+                environment.rmMachineMeta);
+            if (copyExpCodeDirPromise !== undefined) {
+                await copyExpCodeDirPromise;
+            }
+            this.allocateExecutorManagerForEnvironment(environment);
+            const executor = await this.getExecutor(environment.id);
+            environment.runnerWorkingFolder = 
+                executor.joinPath(executor.getRemoteExperimentRootDir(getExperimentId()), 
+                'envs', environment.id)
+            environment.command = `cd ${environment.runnerWorkingFolder} && \
+${environment.command} --job_pid_file ${environment.runnerWorkingFolder}/pid \
+&& echo $? \`date +%s%3N\` >${environment.runnerWorkingFolder}/code`;
+
+            await this.launchEnvironmentOnScheduledMachine(environment);
+
+            environment.status = 'RUNNING';
+            environment.trackingUrl = `file://${rmMachineMeta.ip}:${environment.runnerWorkingFolder}`;
+            this.environmentJobsMap.set(environment.id, environment);
+            deferred.resolve(true);
+        }
+
+        return deferred.promise;
+    }
+
+    /**
+     * give environment an executor
+     * @param environment RemoteMachineEnvironmentDetail
+     */
+    private allocateExecutorManagerForEnvironment(environment: RemoteMachineEnvironmentInformation): void {
+        if (environment.rmMachineMeta === undefined) {
+            throw new Error(`rmMeta not set in trial ${environment.id}`);
+        }
+        const executorManager: ExecutorManager | undefined = this.machineExecutorManagerMap.get(environment.rmMachineMeta);
+        if (executorManager === undefined) {
+            throw new Error(`executorManager not initialized`);
+        }
+        this.environmentExecutorManagerMap.set(environment.id, executorManager);
+    }
+
+    private async getExecutor(environmentId: string): Promise<ShellExecutor> {
+        const executorManager = this.environmentExecutorManagerMap.get(environmentId);
+        if (executorManager === undefined) {
+            throw new Error(`ExecutorManager is not assigned for environment ${environmentId}`);
+        }
+        return await executorManager.getExecutor(environmentId);
+    }
+
+    public async stopEnvironment(environment: EnvironmentInformation): Promise<void> {
+        const executor = await this.getExecutor(environment.id);
+
+        if (environment.status === 'UNKNOWN') {
+            environment.status = 'USER_CANCELED';
+            this.releaseTrialResource(environment);
+            return
+        }
+
+        const jobpidPath: string = `${environment.runnerWorkingFolder}/pid`;
+        try {
+            await executor.killChildProcesses(jobpidPath);
+            this.releaseTrialResource(environment);
+        } catch (error) {
+            this.log.error(`remoteTrainingService.cancelTrialJob: ${error}`);
+        }
+    }
+
+    private async launchEnvironmentOnScheduledMachine(environment: RemoteMachineEnvironmentInformation): Promise<void> {
```

Comment:
the name is strange. how about prepareEnvironment or launchTrialRunner?

Reply:
prepareEnvironment is already used, this function is used to launcher an environment process, changed to use `launcherEnvironment()`